nexus/db-queries/src/db/datastore/db_metadata.rs

```diff
@@ -108,7 +109,273 @@ fn skippable_version(
+    pub fn action(&self) -> &DatastoreSetupAction {
+        &self.action
+    }
+
+    pub fn desired_version(&self) -> &Version {
+        &self.desired
+    }
+}
+
+impl DatastoreSetupAction {
+    // Interprets the combination of access and status to decide what action
+    // should be taken.
+    fn new(access: NexusAccess, status: SchemaStatus) -> Self {
+        use NexusAccess::*;
+        use SchemaStatus::*;
+
+        match (access, status) {
+            // Nexus has been explicitly locked-out of using the database
+            (LockedOut, _) => Self::Refuse,
+
+            // The schema updated beyond what we want, do not use it.
+            (_, NewerThanDesired) => Self::Refuse,
+
+            // If we don't have access yet, but could do something once handoff
+            // occurs, then handoff is needed
+            (
+                DoesNotHaveAccessYet { nexus_id },
+                UpToDate | OlderThanDesired | OlderThanDesiredSkipAccessCheck,
+            ) => Self::NeedsHandoff { nexus_id },
+
+            // This is the most "normal" case: Nexus should have access to the
+            // database, and the schema matches what it wants.
+            (HasExplicitAccess | HasImplicitAccess, UpToDate) => Self::Ready,
+
+            // If this Nexus is allowed to access the schema, but it looks
+            // older than what we expect, we'll need to update the schema to
+            // use it.
+            (
+                HasExplicitAccess | HasImplicitAccess,
+                OlderThanDesired | OlderThanDesiredSkipAccessCheck,
+            ) => Self::Update,
+        }
+    }
+}
+
 impl DataStore {
+    // Checks if the specified Nexus has access to the database.
+    async fn check_nexus_access(
+        &self,
+        nexus_id: OmicronZoneUuid,
+    ) -> Result<NexusAccess, anyhow::Error> {
+        // Check if any "db_metadata_nexus" rows exist.
+        // If they don't exist, treat the database as having access.
+        //
+        // This handles the case for:
+        // - Fresh deployments where RSS hasn't populated the table yet (we need
+        // access to finish "rack_initialization").
+        // - Systems that haven't been migrated to include nexus access control
+        // (we need access to the database to backfill these records).
+        //
+        // After initialization/migration, this conditional should never trigger
+        // again.
+        let any_records_exist = self.database_nexus_access_any_exist().await?;
+        if !any_records_exist {
+            warn!(
+                &self.log,
+                "No db_metadata_nexus records exist - skipping access check";
+                "nexus_id" => ?nexus_id,
+                "explanation" => "This is expected during initial deployment \
+                                  or before migration"
+            );
+            return Ok(NexusAccess::HasImplicitAccess);
+        }
+
+        // Records exist, so enforce the access control check
```

Comment:
```suggestion
        // Records exist, so enforce the identity check
```

We didn't really talk about this but I've been starting to use the term "identity check" rather than "access control" to avoid confusing it with IAM/RBAC/authz sort of stuff.  What do you think?

Reply:
Sounds good to me; definitely reasonable to avoid "access control" as a term because that's overloaded.

Updated in 6034c51c